// @ocubist/error-alchemy/src/validation/parsers/parse/parse.ts
import { ZodError, ZodSchema, z } from "zod";
import { useErrorAlchemy } from "../../../crafting/useErrorAlchemy/useErrorAlchemy";
import { extractZodErrorValidationDetails } from "../../../utility/PropsValidationError";

const { craftMysticError, craftErrorTransmuter } = useErrorAlchemy(
  "ErrorAlchemy",
  "parse-function"
);

export const ParseFailedError = craftMysticError({
  name: "ParseFailedError",
  errorCode: "VALIDATION_ERROR",
  severity: "unexpected",
});

const parseFailedTransmuter = craftErrorTransmuter(
  (err) => err instanceof ZodError,
  (err: ZodError) => {
    const details = extractZodErrorValidationDetails(err);
    return new ParseFailedError({
      message: details.map((d) => d.msg).join(" | "),
      payload: {
        validationErrorDetails: details,
      },
      origin: err,
    });
  }
);

export const parse: <T>(value: T, schema: ZodSchema<T>) => T = (
  value,
  schema
) => {
  try {
    return schema.parse(value);
  } catch (err) {
    throw parseFailedTransmuter.execute(err);
  }
};
