// @ocubist/error-alchemy/dist/crafting/Transmuter/types.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transmuter = exports.TransmuterFunction = exports.DetectorFunction = void 0;
var zod_1 = __importDefault(require("zod"));
var MysticError_1 = require("../../transmuted-errors/MysticError");
var SynthesizedError_1 = require("../../transmuted-errors/SynthesizedError");
exports.DetectorFunction = zod_1.default.function(zod_1.default.tuple([zod_1.default.unknown()]), zod_1.default.boolean());
exports.TransmuterFunction = zod_1.default.function(zod_1.default.tuple([zod_1.default.unknown()]), zod_1.default.union([zod_1.default.instanceof(SynthesizedError_1.SynthesizedError), zod_1.default.instanceof(MysticError_1.MysticError)]));
exports.Transmuter = zod_1.default.object({
    detect: exports.DetectorFunction,
    transmute: exports.TransmuterFunction,
    execute: zod_1.default.function(zod_1.default.tuple([zod_1.default.unknown()]), zod_1.default.unknown()),
});
