// @ocubist/error-alchemy/dist/validation/asserters/assert/assert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assert = exports.AssertFailedError = void 0;
var zod_1 = require("zod");
var useErrorAlchemy_1 = require("../../../crafting/useErrorAlchemy/useErrorAlchemy");
var PropsValidationError_1 = require("../../../utility/PropsValidationError");
var _a = (0, useErrorAlchemy_1.useErrorAlchemy)("ErrorAlchemy", "assert-function"), craftMysticError = _a.craftMysticError, craftErrorTransmuter = _a.craftErrorTransmuter;
exports.AssertFailedError = craftMysticError({
    name: "AssertFailedError",
    errorCode: "DATA_INTEGRITY_VIOLATION",
    severity: "critical",
    cause: "Zod validation failed",
});
var assertFailedTransmuter = craftErrorTransmuter(function (err) {
    // console.log("ENTER DETECTOR");
    return err instanceof zod_1.ZodError;
}, function (err) {
    var details = (0, PropsValidationError_1.extractZodErrorValidationDetails)(err);
    // console.log("ENTER TRANSMUTER");
    return new exports.AssertFailedError({
        message: details.map(function (d) { return d.msg; }).join(" | "),
        payload: {
            validationErrorDetails: details,
        },
        origin: err,
    });
});
function assert(value, schema) {
    // console.log("enters assert function...");
    if (schema !== undefined) {
        try {
            schema.parse(value);
        }
        catch (err) {
            // console.log("The error thrown:", err);
            throw assertFailedTransmuter.execute(err);
        }
    }
}
exports.assert = assert;
